SignalServiceKit/src/Loki/Protocol/Friend Requests/FriendRequestProtocol.swift

```diff
@@ -40,55 +62,126 @@ public final class FriendRequestProtocol : NSObject {
+            friendRequestStatuses = linkedDevices.map { device in
+                return storage.getFriendRequestStatus(for: device, transaction: transaction)
+            }
+        }
+
+        if friendRequestStatuses.contains(where: { $0 == .friends }) { return .friends }
+        if friendRequestStatuses.contains(where: { $0 == .requestReceived }) { return .received }
+        if friendRequestStatuses.contains(where: { $0 == .requestSent }) { return .sent }
+
+        return .none
+    }
+
     // MARK: - Sending
-    @objc(acceptFriendRequestFrom:in:using:)
-    public static func acceptFriendRequest(from hexEncodedPublicKey: String, in thread: TSThread, using transaction: YapDatabaseReadWriteTransaction) {
+    @objc(acceptFriendRequestFromHexEncodedPublicKey:using:)
+    public static func acceptFriendRequest(from hexEncodedPublicKey: String, using transaction: YapDatabaseReadWriteTransaction) {
         // Accept all outstanding friend requests associated with this user and try to establish sessions with the
         // subset of their devices that haven't sent a friend request.
-        let linkedDeviceThreads = LokiDatabaseUtilities.getLinkedDeviceThreads(for: hexEncodedPublicKey, in: transaction) // This doesn't create new threads if they don't exist yet
-        // FIXME: Capture send failures
-        for thread in linkedDeviceThreads {
-            if thread.hasPendingFriendRequest {
-                sendFriendRequestAcceptanceMessage(to: thread.contactIdentifier(), in: thread, using: transaction) // NOT hexEncodedPublicKey
-                thread.saveFriendRequestStatus(.friends, with: transaction)
-            } else {
-                MultiDeviceProtocol.getAutoGeneratedMultiDeviceFRMessageSend(for: thread.contactIdentifier(), in: transaction) // NOT hexEncodedPublicKey
-                .done(on: OWSDispatch.sendingQueue()) { autoGeneratedFRMessageSend in
-                    let messageSender = SSKEnvironment.shared.messageSender
-                    messageSender.sendMessage(autoGeneratedFRMessageSend)
+        guard ECKeyPair.isValidHexEncodedPublicKey(candidate: hexEncodedPublicKey) else {
+            assertionFailure("Invalid session ID \(hexEncodedPublicKey)")
+            return;
+        }
+
+        let linkedDevices = LokiDatabaseUtilities.getLinkedDeviceHexEncodedPublicKeys(for: hexEncodedPublicKey, in: transaction)
+        for device in linkedDevices {
+            let friendRequestStatus = storage.getFriendRequestStatus(for: device, transaction: transaction)
+            if friendRequestStatus == .requestReceived {
+                storage.setFriendRequestStatus(.friends, for: device, transaction: transaction)
+                sendFriendRequestAcceptanceMessage(to: device, using: transaction)
+            } else if friendRequestStatus == .requestSent {
+                // We sent a friend request to this device before, how can we be sure that it hasn't expired?
```

Comment:
Shouldn't the status then be `.expired`?

Reply:
No, I just put this status there so I don't forget about the comment